File: VC_HIL/python/inverterhilgui/fake_target_backend.py
"""Faithful in-memory backend for host tests; contains no MATLAB dependency."""

from .parameter_contract import parameterContract


class FakeTargetBackend:
    def __init__(self, available_paths=None):
        if not available_paths:
            available_paths = [entry.structPath for entry in parameterContract()]
        self.AvailablePaths = list(available_paths)
        self.Values = {path: 0 for path in self.AvailablePaths}
        self.Signals = {}
        self.FailNextCall = False
        self.AppliedOverride = None
        self.WriteCount = 0
        self.WriteLog = []
        self.InstrumentCount = 0
        self.Connected = False
        self.Loaded = False
        self.Running = False
        self.Name = "FakeTarget"
        self.ExecutionTimeS = 0.0
        self.CurrentApplication = ""

    def _fail_if_requested(self, action):
        if self.FailNextCall:
            self.FailNextCall = False
            raise RuntimeError(f"The fake target was told to fail {action}.")

    def _require_connected(self):
        if not self.Connected:
            raise RuntimeError("The fake target is not connected.")

    def connect(self):
        self._fail_if_requested("connect")
        self.Connected = True

    def disconnect(self):
        self.Connected = self.Loaded = self.Running = False

    def load(self, applicationName):
        self._fail_if_requested("load")
        self._require_connected()
        self.Loaded = True
        self.Running = False
        self.CurrentApplication = "inverter_hil"

    def start(self):
        self._fail_if_requested("start")
        self._require_connected()
        if not self.Loaded:
            raise RuntimeError("No application is loaded on the fake target.")
        self.Running = True

    def stop(self):
        self._fail_if_requested("stop")
        self.Running = False

    def reset(self):
        self._fail_if_requested("reset")
        self.Running = False
        self.ExecutionTimeS = 0.0

    def getparam(self, path):
        self._require_connected()
        if path not in self.Values:
            raise KeyError(f"The fake target does not expose {path}.")
        return self.Values[path]

    def setparam(self, path, value):
        self._fail_if_requested("setparam")
        self._require_connected()
        if path not in self.Values:
            raise KeyError(f"The fake target does not expose {path}.")
        self.Values[path] = value if self.AppliedOverride is None else self.AppliedOverride
        self.WriteCount += 1
        self.WriteLog.append((path, value))

    def getsignal(self, blockPath, port):
        self._require_connected()
        return self.Signals.get((blockPath, port), 0)

    def isConnected(self):
        return self.Connected

    def applicationState(self):
        if not self.Connected:
            return "disconnected"
        if self.Running:
            return "running"
        if self.Loaded:
            return "loaded"
        return "connected"

    def currentApplicationName(self):
        return self.CurrentApplication

    def executionTimeS(self):
        return self.ExecutionTimeS

    def targetName(self):
        return self.Name

    def availableParameters(self):
        return list(self.AvailablePaths)

    def addInstrument(self, instrument):
        self.InstrumentCount += 1

    def removeInstruments(self):
        self.InstrumentCount = 0

File: VC_HIL/inverter/+inverterhilgui/+sg_adapters/fakeTargetBackend.m
classdef fakeTargetBackend < handle
    %FAKETARGETBACKEND In-memory test double for the Simulink Real-Time seam.
    %
    %   This class exists ONLY so that INVERTERHILGUI.TARGETSESSION can be
    %   unit-tested without a Speedgoat. It is NOT a demo mode and it is NEVER
    %   wired into INVERTER_HIL_APP: the app constructs a target session with
    %   no injected backend, which yields the real Simulink Real-Time backend
    %   and a hard failure when no target is present.
    %
    %   It implements the backend interface expected by TARGETSESSION:
    %     connect, disconnect, load, start, stop, reset, getparam, setparam,
    %     getsignal, isConnected, applicationState, executionTimeS,
    %     targetName, availableParameters, addInstrument, removeInstruments.

    properties
        % Parameter paths this fake application claims to expose.
        AvailablePaths = {}
        % Values keyed by parameter path.
        Values
        % When true, the next write or lifecycle call fails.
        FailNextCall = false
        % Value the target reports back regardless of what was written.
        AppliedOverride
        % Number of writes performed.
        WriteCount = 0
        % Ordered log of {path, value} pairs actually written.
        WriteLog = {}
        % Number of times an instrument was registered.
        InstrumentCount = 0
    end

    properties (SetAccess = private)
        Connected = false
        Loaded = false
        Running = false
        Name = 'FakeTarget'
        ExecutionTimeS = 0
        CurrentApplication = ''
    end

    methods
        function obj = fakeTargetBackend(availablePaths)
            %FAKETARGETBACKEND Construct a fake backend with known parameters.
            obj.Values = containers.Map('KeyType', 'char', ...
                'ValueType', 'any');
            if nargin < 1 || isempty(availablePaths)
                contract = inverterhilgui.params.parameterContract();
                availablePaths = cell(1, numel(contract));
                for index = 1:numel(contract)
                    availablePaths{index} = contract(index).structPath;
                end
            end
            obj.AvailablePaths = availablePaths;
            for index = 1:numel(obj.AvailablePaths)
                obj.Values(obj.AvailablePaths{index}) = 0;
            end
        end

        function connect(obj)
            obj.failIfRequested('connect');
            obj.Connected = true;
        end

        function disconnect(obj)
            obj.Connected = false;
            obj.Loaded = false;
            obj.Running = false;
        end

        function load(obj, ~)
            obj.failIfRequested('load');
            obj.requireConnected();
            obj.Loaded = true;
            obj.Running = false;
            obj.CurrentApplication = 'inverter_hil';
        end

        function start(obj)
            obj.failIfRequested('start');
            obj.requireConnected();
            if ~obj.Loaded
                error('inverterhilgui:FakeTargetNotLoaded', ...
                    'No application is loaded on the fake target.');
            end
            obj.Running = true;
        end

        function stop(obj)
            obj.failIfRequested('stop');
            obj.Running = false;
        end

        function name = currentApplicationName(obj)
            name = obj.CurrentApplication;
        end

        function reset(obj)
            obj.failIfRequested('reset');
            obj.Running = false;
            obj.ExecutionTimeS = 0;
        end

        function value = getparam(obj, path)
            obj.requireConnected();
            if ~obj.Values.isKey(path)
                error('inverterhilgui:FakeTargetUnknownParameter', ...
                    'The fake target does not expose %s.', path);
            end
            value = obj.Values(path);
        end

        function setparam(obj, path, value)
            obj.failIfRequested('setparam');
            obj.requireConnected();
            if ~obj.Values.isKey(path)
                error('inverterhilgui:FakeTargetUnknownParameter', ...
                    'The fake target does not expose %s.', path);
            end
            if ~isempty(obj.AppliedOverride)
                obj.Values(path) = obj.AppliedOverride;
            else
                obj.Values(path) = value;
            end
            obj.WriteCount = obj.WriteCount + 1;
            obj.WriteLog{end + 1, 1} = {path, value};
        end

        function value = getsignal(obj, ~, ~)
            % No fake test currently exercises READLIVEIO; this exists only
            % so FAKETARGETBACKEND keeps satisfying the backend interface.
            obj.requireConnected();
            value = 0;
        end

        function value = isConnected(obj)
            value = obj.Connected;
        end

        function state = applicationState(obj)
            if ~obj.Connected
                state = 'disconnected';
            elseif obj.Running
                state = 'running';
            elseif obj.Loaded
                state = 'loaded';
            else
                state = 'connected';
            end
        end

        function value = executionTimeS(obj)
            value = obj.ExecutionTimeS;
        end

        function value = targetName(obj)
            value = obj.Name;
        end

        function paths = availableParameters(obj)
            paths = obj.AvailablePaths;
        end

        function addInstrument(obj, ~)
            obj.InstrumentCount = obj.InstrumentCount + 1;
        end

        function removeInstruments(obj)
            obj.InstrumentCount = 0;
        end
    end

    methods (Access = private)
        function failIfRequested(obj, action)
            if obj.FailNextCall
                obj.FailNextCall = false;
                error('inverterhilgui:FakeTargetFailure', ...
                    'The fake target was told to fail %s.', action);
            end
        end

        function requireConnected(obj)
            if ~obj.Connected
                error('inverterhilgui:FakeTargetNotConnected', ...
                    'The fake target is not connected.');
            end
        end
    end
end
